src/sandstorm/run-bundle.c++

```diff
@@ -1861,17 +1873,55 @@ private:
       return false;
     }
 
-    // Start http request to download bundle.
+    // Download bundle to temporary file.
     auto url = kj::str("https://dl.sandstorm.io/sandstorm-", targetBuild, ".tar.xz");
+    auto file = openTemporary("/var/tmp/sandstorm-update");
     context.warning(kj::str("Downloading: ", url));
-    auto download = kj::heap<CurlRequest>(url);
-    int fd = download->getPipe();
-    unpackUpdate(fd, kj::mv(download), targetBuild);
+    CurlRequest(url, file);
+    KJ_SYSCALL(lseek(file, 0, SEEK_SET));
+
+    // Verify signature.
+    {
+      context.warning("Checking signature...");
+      KJ_ON_SCOPE_FAILURE(
+          context.warning("*** SIGNATURE CHECK FAILED: please notify security@sandstorm.io ***"));
```

Comment:
Note that the message will be followed by a more detailed (but technical) exception.

I suppose I could move the dire bits into the specific error messages below for a failed signature. OTOH, an attacker who can serve a bogus package would presumably block the signature from being downloaded rather than let it be downloaded and proved invalid. So maybe what I really need to do is have this message say: "This may be a temporary glitch, but if the problem persists, it may indicate an attack. Please notify security@sandstorm.io."

Of course given that a signature check is taking place and therefore an attacker would gain nothing from serving a bogus package, maybe it's safe to assume that the problem almost certainly isn't an attack. If the the attacker were trying to block you from updating, they'd block all communication, so the package would never have downloaded in the first place.

Hrm.